bellman/src/groth16/generator/assembly/eval.rs

```diff
@@ -0,0 +1,205 @@
+    ext: &mut [E::G1],
+
+    // Inverse coefficient for ext elements
+    inv: &E::Fr,
+
+    // Trapdoors
+    alpha: &E::Fr,
+    beta: &E::Fr,
+
+    // Worker
+    worker: &Worker,
+) {
+    // Sanity check
+    assert_eq!(a.len(), at.len());
+    assert_eq!(a.len(), bt.len());
+    assert_eq!(a.len(), ct.len());
+    assert_eq!(a.len(), b_g1.len());
+    assert_eq!(a.len(), b_g2.len());
+    assert_eq!(a.len(), ext.len());
+
+    // Evaluate polynomials in multiple threads
+    worker.scope(a.len(), |scope, chunk| {
+        for ((((((a, b_g1), b_g2), ext), at), bt), ct) in a
+            .chunks_mut(chunk)
+            .zip(b_g1.chunks_mut(chunk))
+            .zip(b_g2.chunks_mut(chunk))
+            .zip(ext.chunks_mut(chunk))
+            .zip(at.chunks(chunk))
+            .zip(bt.chunks(chunk))
+            .zip(ct.chunks(chunk))
+        {
+            let mut g1_wnaf = g1_wnaf.shared();
+            let mut g2_wnaf = g2_wnaf.shared();
+
+            scope.spawn(move || {
+                for ((((((a, b_g1), b_g2), ext), at), bt), ct) in a
+                    .iter_mut()
+                    .zip(b_g1.iter_mut())
+                    .zip(b_g2.iter_mut())
+                    .zip(ext.iter_mut())
+                    .zip(at.iter())
+                    .zip(bt.iter())
+                    .zip(ct.iter())
+                {
+                    fn eval_at_tau<E: Engine>(
+                        powers_of_tau: &[Scalar<E>],
+                        p: &[(E::Fr, usize)],
+                    ) -> E::Fr {
+                        let mut acc = E::Fr::zero();
+
+                        for &(ref coeff, index) in p {
+                            let mut n = powers_of_tau[index].0;
+                            n.mul_assign(coeff);
+                            acc.add_assign(&n);
+                        }
+
+                        acc
+                    }
+
+                    // Evaluate QAP polynomials at tau
+                    let mut at = eval_at_tau(powers_of_tau, at);
+                    let mut bt = eval_at_tau(powers_of_tau, bt);
+                    let ct = eval_at_tau(powers_of_tau, ct);
+
+                    // Compute A query (in G1)
+                    if !at.is_zero() {
+                        *a = g1_wnaf.scalar(at.into_repr());
+                    }
+
+                    // Compute B query (in G1/G2)
+                    if !bt.is_zero() {
+                        let bt_repr = bt.into_repr();
+                        *b_g1 = g1_wnaf.scalar(bt_repr);
+                        *b_g2 = g2_wnaf.scalar(bt_repr);
+                    }
+
+                    at.mul_assign(&beta);
+                    bt.mul_assign(&alpha);
+
+                    let mut e = at;
+                    e.add_assign(&bt);
+                    e.add_assign(&ct);
+                    e.mul_assign(inv);
+
+                    *ext = g1_wnaf.scalar(e.into_repr());
+                }
+
+                // Batch normalize
```

Comment:
The principle I generally try and follow with functions such as these are:
1) Don't mix layers of abstraction, e.g. sanity checking (which I personally don't like as asserts unless they're caught) and higher order function calls
2) Try and keep functions such as eval as close to natural language as possible. e.g. just function calls and comments

The reason being that these higher order functions should be easily read to work out what the rest of a class or grouping of code does.

Reply:
Eval is still on the todo list for refactoring. I haven’t changed the function except to remove the worker pattern.

Reply:
Has been refactored. Note to rereview.